format/macho/macho.go

```diff
@@ -0,0 +1,835 @@
+					d.FieldU32("align")
+					d.FieldU32("reloff")
+					d.FieldU32("nreloc")
+					// get section type
+					sectionType := d.U32()
+					sectionTypeMasked := sectionType & 0xFF
+					d.FieldValueStr("type", sectionTypes[sectionTypeMasked])
+					// rewind 32 bits and parse flags
+					d.SeekRel(-4 * 8)
+					d.FieldStruct("flags", parseFlags(sectionFlags))
+					d.FieldU32("reserved1")
+					d.FieldU32("reserved2")
+					if archBits == 64 {
+						d.FieldU32("reserved3")
+					}
+					nsectIdx++
+				})
+		case LC_TWOLEVEL_HINTS:
+			d.FieldU32("offset")
+			d.FieldU32("nhints")
+		case LC_LOAD_DYLIB, LC_ID_DYLIB, LC_LOAD_UPWARD_DYLIB, LC_LOAD_WEAK_DYLIB, LC_LAZY_LOAD_DYLIB, LC_REEXPORT_DYLIB:
+			d.FieldStruct("dylib_command", func(d *decode.D) {
+				offset := d.FieldU32("offset")
+				d.FieldU32("timestamp", timestampMapper)
+				d.FieldU32("current_version")
+				d.FieldU32("compatibility_version")
+				d.FieldUTF8NullFixedLen("name", int(cmdsize)-int(offset))
+			})
+		case LC_LOAD_DYLINKER, LC_ID_DYLINKER, LC_DYLD_ENVIRONMENT:
+			offset := d.FieldU32("offset")
+			d.FieldUTF8NullFixedLen("name", int(cmdsize)-int(offset))
+		case LC_RPATH:
+			offset := d.FieldU32("offset")
+			d.FieldUTF8NullFixedLen("name", int(cmdsize)-int(offset))
+		case LC_PREBOUND_DYLIB:
+			// https://github.com/aidansteele/osx-abi-macho-file-format-reference#prebound_dylib_command
+			d.U32() // name_offset
+			nmodules := d.FieldU32("nmodules")
+			d.U32() // linked_modules_offset
+			d.FieldUTF8Null("name")
+			d.FieldBitBufFn("linked_modules", func(d *decode.D) *bitio.Buffer {
+				return d.RawLen(int64((nmodules / 8) + (nmodules % 8)))
+			})
+		case LC_THREAD, LC_UNIXTHREAD:
+			d.FieldU32("flavor")
+			count := d.FieldU32("count")
+			d.FieldStruct("state", func(d *decode.D) {
+				switch cpuType {
+				case 0x7:
+					threadStateI386Decode(d)
+				case 0xC:
+					threadStateARM32Decode(d)
+				case 0x13:
+					threadStatePPC32Decode(d)
+				case 0x1000007:
+					threadStateX8664Decode(d)
+				case 0x100000C:
+					threadStateARM64Decode(d)
+				case 0x1000013:
+					threadStatePPC64Decode(d)
+				default:
+					d.FieldRawLen("state", int64(count*32))
+				}
+			})
+		case LC_ROUTINES, LC_ROUTINES_64:
+			if archBits == 32 {
+				d.FieldU32("init_address", scalar.Hex)
+				d.FieldU32("init_module")
+				d.FieldU32("reserved1", d.BitBufIsZero())
+				d.FieldU32("reserved2", d.BitBufIsZero())
+				d.FieldU32("reserved3", d.BitBufIsZero())
+				d.FieldU32("reserved4", d.BitBufIsZero())
+				d.FieldU32("reserved5", d.BitBufIsZero())
+				d.FieldU32("reserved6", d.BitBufIsZero())
+			} else {
+				d.FieldU64("init_address", scalar.Hex)
+				d.FieldU64("init_module")
+				d.FieldU64("reserved1", d.BitBufIsZero())
+				d.FieldU64("reserved2", d.BitBufIsZero())
+				d.FieldU64("reserved3", d.BitBufIsZero())
+				d.FieldU64("reserved4", d.BitBufIsZero())
+				d.FieldU64("reserved5", d.BitBufIsZero())
+				d.FieldU64("reserved6", d.BitBufIsZero())
```

Comment:
BitBufIsZero only works if the actual value is a buffer, so this is probably a nop. Maybe it should even panic if used with wrong type